// ginlibs-type-check/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function isString(v) {
  return typeof v === 'string';
}
function isFinite(v) {
  return Number.isFinite(v);
}
function isNumber(v) {
  return typeof v === 'number' && !Number.isNaN(v) && isFinite(v);
}
function isNull(v) {
  return v === null;
}
function isUndefined(v) {
  return v === undefined;
}
function isBoolean(v) {
  return v === true || v === false;
}
function isObject(v) {
  return isThisType(v, 'Object');
}
function isEmptyObject(v) {
  return isThisType(v, 'Object') && Object.keys(v).length <= 0;
}
function isArray(v) {
  return Array.isArray(v);
}
function isEmptyArray(v) {
  return Array.isArray(v) && v.length <= 0;
}
function isNilVal(v, ...args) {
  args.push(v);
  return args.every(it => isUndefined(it) || isNull(it));
}
function isFunc(v) {
  return typeof v === 'function';
}
function isPromise(v) {
  return v && isFunc(v.then);
}
function isDate(v) {
  return isThisType(v, 'Date') && isNumber(v.getTime());
}
function isThisType(val, type) {
  const valType = Object.prototype.toString.call(val);
  const trueType = valType.slice(8, valType.length - 1);
  return type === trueType;
}

exports.isArray = isArray;
exports.isBoolean = isBoolean;
exports.isDate = isDate;
exports.isEmptyArray = isEmptyArray;
exports.isEmptyObject = isEmptyObject;
exports.isFinite = isFinite;
exports.isFunc = isFunc;
exports.isNilVal = isNilVal;
exports.isNull = isNull;
exports.isNumber = isNumber;
exports.isObject = isObject;
exports.isPromise = isPromise;
exports.isString = isString;
exports.isThisType = isThisType;
exports.isUndefined = isUndefined;
